server: new discrete security overrides to replace `--insecure`

knz opened the pull request:
Fixes #74329.

tldr: This change breaks down the behavior of `--insecure` into a set of discrete security feature flags that can be applied separately. This makes it possible for a user to opt out from a certain protection without opting out from every protection.

An example of this is the new ability to disable the TLS protection on the HTTP interface while retaining authentication.

The new feature flags can be controlled via the command line flag `--security-overrides`, which takes a set of overrides delimited by commas.

This change also folds existing command-line flag under the new mechanism. For example:
- `--allow-sql-without-tls` (an experimental, previously hidden flag) is now `--security-overrides=disable-sql-require-tls`
- `--unencrypted-localhost-http` is now `--http-addr=localhost --security-overrides=disable-http-tls`

Here is a summary of the newly introduced feature flags:

| Flag                                | Description                                                                                                                         |
|-------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------|
| `disable-tls`                       | Disable TLS on all interfaces.                                                                                                      |
| `disable-rpc-tls`                   | Disable TLS for RPC connections. Implies `disable-remote-certs-retrieval`. Currently implies `disable-rpc-authn`.                   |
| `disable-rpc-authn`                 | Disable authentication and authorization for RPC connections. Implies `disable-remote-certs-retrieval`.                             |
| `disable-sql-authn`                 | Disable authentication for SQL client connections. (As if the method `trust` was applied to all connections.)                       |
| `disable-sql-tls`                   | Disable TLS for SQL connections. Implies `disable-sql-require-tls`.                                                                 |
| `disable-sql-require-tls`           | Disable the requirement to use TLS for SQL clients. TLS is still allowed but becomes optional.  Replaces `--allow-sql-without-tls`. |
| `disable-sql-set-credentials`       | Disable the ability to set user passwords in SQL.                                                                                   |
| `disable-http-tls`                  | Disable TLS for HTTP clients. Implies `disable-remote-certs-retrieval`.    Replaces `--unencrypted-localhost-http`.                 |
| `disable-http-authn`                | Disable HTTP authentication. Users do not need to log in to perform HTTP queries.  Implies `disable-remote-certs-retrieval`.        |
| `disable-remote-certs-retrieval`    | Disable the retrieval of TLS certificates using HTTP and RPC requests.                                                              |
| `disable-cluster-name-verification` | Disable the RPC-level cluster name verification protocol.  Replaces `--disable-cluster-name-verification`.                          |
| `disable-all`                       | Implies all of the above.                                                                                                           |




See the individual commits for further details.

knz committed:
base: new type `SecurityOverrides`

We'll want to use this as a more fine-grained mechanism to replace
the "insecure" boolean.

Release note (cli change): A new flag `--security-overrides` is
introduced to provide more fine grained control over security
protections in replacement to `--insecure`.

knz committed:
sql: make setting passwords subject to a discrete security override

Release note (sql change): Whether SQL clients can modify passwords
with ALTER/CREATE USER/ROLE WITH PASSWORD is now configurable via the
security overrides `disable-sql-set-credentials`, in addition to the
coarse grained flag `--insecure`.

knz committed:
cli,rpc: replace the discrete bool DisableClusterNameVerification by a sec override

Release note (cli change): The command-line flag
`--disable-cluster-name-verification` (needed when changing/setting
the value of `--cluster-name`) is now deprecated, replaced
by a new security override flag that can be set via
`--security-overrides=disable-cluster-name-verification`.
The previous command-line flag is now an alias for setting this override.

knz committed:
cli,sql: replace the discrete bool AcceptSQLWithoutTLS by a sec override

Release note (cli change): The experimental command-line flag
`--accept-sql-without-tls` (which enables, but does not require, SQL
clients to omit TLS to connect to a CockroachDB server) is now
deprecated, replaced by a new security override flag that can be set
via `--security-overrides=disable-sql-require-tls`.  The previous
command-line flag is now an alias for setting this override.

knz committed:
server,cli: introduce a new discrete flag for HTTP TLS

Release note (cli change): It is now possible to disable TLS
piecemeal on the HTTP interface, without the requirement
that the HTTP interface listens on `localhost` like
`--unencrypted-localhost-http` implies. This can be set
using the new override flag `--security-overrides=disable-http-tls`.

knz committed:
server: block remote cert retrieval under new sec override

Release note (cli change): The ability to retrieve TLS certificates
under an API is now controlled by the security override
`disable-remote-certs-retrieval`. It is automatically implied by
`--insecure` and any of the other security overrides
`disable-rpc-tls`, `disable-rpc-authn`, `disable-http-tls` or
`disable-http-authn`, but can also be set piecemeal.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74330)
<!-- Reviewable:end -->

knz committed:
base,server: replace the discrete bool EnableWebSessionAuthentication by a sec override

Release note (cli change): It is now possible to disable
authentication on the web UI piecemeal using the new security override
`disable-http-authn`. Support for the environment variable
`COCKROACH_DISABLE_WEB_LOGIN` is preserved, although that mechanism
might be removed in a later version of CockroachDB.

knz committed:
pgwire: make SQL authentication controllable via a security override

Release note (cli change): It is now possible to disable SQL
authentication (i.e. override the authentication method to always
be `trust`) for a single CockroachDB node using the security
override `disable-sql-authn`.

knz committed:
pgwire: make SQL TLS controllable via a security override

Release note (cli change): The availability of TLS for SQL client
connections over TCP is now controllable via the security override
`disable-sql-tls`.
